test/helper/log_and_filtering.bash

```diff
@@ -0,0 +1,100 @@
+#!/bin/bash
+
+# ? ABOUT: Functions defined here aid in working with logs and filtering them.
+
+# ! ATTENTION: This file is loaded by `common.sh` - do not load it yourself!
+# ! ATTENTION: This file requires helper functions from `common.sh`!
+
+# shellcheck disable=SC2034,SC2155
+
+# Assert that the number of lines output by a previous command matches the given amount (${1}).
+# `lines` is a special BATS variable updated via `run`.
+#
+# @param ${1} = number of lines that the output should have
+function _should_output_number_of_lines() {
+  # shellcheck disable=SC2154
+  assert_equal "${#lines[@]}" "${1:?Number of lines not provided}"
+}
+
+# Filters a service's logs (under `/var/log/supervisor/<SERVICE>.log`) given a specific string.
+#
+# @param ${1} = service name
+# @param ${2} = string to filter by
+# @param ...  = options given to `grep` (which is used to filter logs)
+function _filter_service_log() {
+  local SERVICE=${1:?Service name must be provided}
+  local STRING=${2:?String to match must be provided}
+  shift 2
+
+  local FILE="/var/log/supervisor/${SERVICE}.log"
+  [[ -f ${FILE} ]] || FILE="/var/log/mail/${SERVICE}.log" # Fallback to alternative log location:
+  _run_in_container grep "${@}" "${STRING}" "${FILE}"
```

Comment:
`${@}` should default/fallback to `--fixed-strings` if possible, or use a separate `_regex` variant method.

The benefit of a `_regex` variant method is that it's easier to grep the test files for usage of where we're actually using that optional parameter, but that's a minor concern, I'm happy regardless of how you want to approach it.

Reply:
Having `_regex` will not be DRY with our helpers at all. But I guess the concern about searching for the method is a good one. I will implement that.